File: src/main/java/top/potmot/jimmercodegen/config/JimmerConfig.java
package top.potmot.jimmercodegen.config;

import org.babyfish.jimmer.sql.runtime.DefaultExecutor;
import org.babyfish.jimmer.sql.runtime.ExecutionPurpose;
import org.babyfish.jimmer.sql.runtime.Executor;
import org.babyfish.jimmer.sql.runtime.ExecutorContext;
import org.babyfish.jimmer.sql.runtime.SqlFunction;
import org.babyfish.jimmer.sql.runtime.StatementFactory;
import org.jetbrains.annotations.Nullable;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import top.potmot.jimmercodegen.utils.LogOutputUtil;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.util.List;

@Configuration
public class JimmerConfig {
    private static final Logger LOGGER = LoggerFactory.getLogger(JimmerConfig.class);

    @Value("${gen.show-sql}")
    public Boolean showSql;

    @Bean
    public Executor executor() {
        return new Executor() {
            @Override
            public <R> R execute(
                    Connection con, String sql, List<Object> variables, ExecutionPurpose purpose, @Nullable ExecutorContext ctx, StatementFactory statementFactory, SqlFunction<PreparedStatement, R> block
            ) {
                long start = System.currentTimeMillis();
                R result = DefaultExecutor
                        .INSTANCE
                        .execute(
                                con,
                                sql,
                                variables,
                                purpose,
                                ctx,
                                statementFactory,
                                block
                        );
                if (showSql) {
                    long cost = System.currentTimeMillis() - start;
                    LogOutputUtil.infoSqlLog(sql, variables, cost, result, LOGGER);
                }
                return result;
            }
        };
    }
}

File: src/main/kotlin/top/potmot/external/config/JimmerConfig.kt
package top.potmot.external.config

import org.babyfish.jimmer.sql.runtime.*
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import top.potmot.external.util.LogUtils
import java.sql.Connection
import java.sql.PreparedStatement

@Configuration
class JimmerConfig {
    @Value("\${gen.show-sql:false}")
    var showSql: Boolean? = null
    @Bean
    fun executor(): Executor {
        return object : Executor {
            override fun <R> execute(
                con: Connection,
                sql: String,
                variables: List<Any?>,
                purpose: ExecutionPurpose,
                ctx: ExecutorContext?,
                statementFactory: StatementFactory?,
                block: SqlFunction<PreparedStatement, R>
            ): R {
                val start = System.currentTimeMillis()
                val result = DefaultExecutor.INSTANCE
                    .execute(
                        con,
                        sql,
                        variables,
                        purpose,
                        ctx,
                        statementFactory,
                        block
                    )
                if (showSql!!) {
                    val cost = System.currentTimeMillis() - start
                    LogUtils.infoSqlLog(sql, variables, cost, result, LOGGER)
                }
                return result
            }
        }
    }

    companion object {
        private val LOGGER = LoggerFactory.getLogger(JimmerConfig::class.java)
    }
}
